src/libraries/System.IO.Hashing/src/System/IO/Hashing/NonCryptographicHashAlgorithm.cs

```diff
@@ -0,0 +1,326 @@
+
+        /// <summary>
+        ///   Appends the contents of <paramref name="source"/> to the data already
+        ///   processed for the current hash computation.
+        /// </summary>
+        /// <param name="source">The data to process.</param>
+        /// <exception cref="ArgumentNullException">
+        ///   <paramref name="source"/> is <see langword="null"/>.
+        /// </exception>
+        public void Append(byte[] source)
+        {
+            if (source is null)
+                throw new ArgumentNullException(nameof(source));
+
+            Append(new ReadOnlySpan<byte>(source));
+        }
+
+        /// <summary>
+        ///   Appends the contents of <paramref name="stream"/> to the data already
+        ///   processed for the current hash computation.
+        /// </summary>
+        /// <param name="stream">The data to process.</param>
+        /// <exception cref="ArgumentNullException">
+        ///   <paramref name="stream"/> is <see langword="null"/>.
+        /// </exception>
+        /// <seealso cref="AppendAsync(Stream, CancellationToken)"/>
+        public void Append(Stream stream)
+        {
+            if (stream is null)
+                throw new ArgumentNullException(nameof(stream));
+
+            byte[] buffer = ArrayPool<byte>.Shared.Rent(4096);
+
+            while (true)
+            {
+                int read = stream.Read(buffer, 0, buffer.Length);
+
+                if (read == 0)
+                {
+                    break;
+                }
+
+                Append(new ReadOnlySpan<byte>(buffer, 0, read));
+            }
+
+            ArrayPool<byte>.Shared.Return(buffer);
+        }
+
+        /// <summary>
+        ///   Asychronously reads the contents of <paramref name="stream"/>
+        ///   and appends them to the data already
+        ///   processed for the current hash computation.
+        /// </summary>
+        /// <param name="stream">The data to process.</param>
+        /// <param name="cancellationToken">
+        ///   The token to monitor for cancellation requests.
+        ///   The default value is <see cref="CancellationToken.None"/>.
+        /// </param>
+        /// <exception cref="ArgumentNullException">
+        ///   <paramref name="stream"/> is <see langword="null"/>.
+        /// </exception>
+        public Task AppendAsync(Stream stream, CancellationToken cancellationToken = default)
+        {
+            if (stream is null)
+                throw new ArgumentNullException(nameof(stream));
+
+            return AppendAsyncCore(stream, cancellationToken);
+        }
+
+        private async Task AppendAsyncCore(Stream stream, CancellationToken cancellationToken)
+        {
+            byte[] buffer = ArrayPool<byte>.Shared.Rent(4096);
+
+            while (true)
+            {
+                int read = await stream.ReadAsync(buffer, 0, buffer.Length, cancellationToken).ConfigureAwait(false);
```

Comment:
This is an example of where it'd be nice if we had a build that could utilize `Stream.ReadAsync(Memory<byte>, ...)`.  Reads will generally be cheaper.